runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/commits/StreamingEngineWorkCommitterTest.java

```diff
@@ -0,0 +1,208 @@
+
+  @Before
+  public void setUp() {
+    fakeWindmillServer =
+        new FakeWindmillServer(
+            errorCollector, ignored -> Optional.of(Mockito.mock(ComputationState.class)));
+    commitWorkStreamFactory =
+        WindmillStreamPool.create(
+                1, Duration.standardMinutes(1), fakeWindmillServer::commitWorkStream)
+            ::getCloseableStream;
+  }
+
+  @After
+  public void cleanUp() {
+    workCommitter.stop();
+  }
+
+  private StreamingEngineWorkCommitter createWorkCommitter(
+      Consumer<CompleteCommit> onCommitComplete) {
+    return StreamingEngineWorkCommitter.create(commitWorkStreamFactory, 1, onCommitComplete);
+  }
+
+  @Test
+  public void testCommit_sendsCommitsToStreamingEngine() {
+    workCommitter = createWorkCommitter(ignored -> {});
+    List<Commit> commits = new ArrayList<>();
+    for (int i = 1; i <= 5; i++) {
+      Work work = createMockWork(i, ignored -> {});
+      WorkItemCommitRequest commitRequest =
+          WorkItemCommitRequest.newBuilder()
+              .setKey(work.getWorkItem().getKey())
+              .setShardingKey(work.getWorkItem().getShardingKey())
+              .setWorkToken(work.getWorkItem().getWorkToken())
+              .setCacheToken(work.getWorkItem().getCacheToken())
+              .build();
+      commits.add(Commit.create(commitRequest, createComputationState("computationId-" + i), work));
+    }
+
+    workCommitter.start();
+    commits.forEach(workCommitter::commit);
+
+    Map<Long, WorkItemCommitRequest> committed =
+        fakeWindmillServer.waitForAndGetCommits(commits.size());
+
+    for (Commit commit : commits) {
+      WorkItemCommitRequest request = committed.get(commit.work().getWorkItem().getWorkToken());
+      assertNotNull(request);
+      assertThat(request).isEqualTo(commit.request());
+    }
+  }
+
+  @Test
+  public void testCommit_handlesFailedCommits() {
+    Set<CompleteCommit> failedCommits = new HashSet<>();
+    workCommitter = createWorkCommitter(failedCommits::add);
+    List<Commit> commits = new ArrayList<>();
+    for (int i = 1; i <= 10; i++) {
+      Work work = createMockWork(i, ignored -> {});
+      // Fail half of the work.
+      if (i % 2 == 0) {
+        work.setFailed();
+      }
+      WorkItemCommitRequest commitRequest =
+          WorkItemCommitRequest.newBuilder()
+              .setKey(work.getWorkItem().getKey())
+              .setShardingKey(work.getWorkItem().getShardingKey())
+              .setWorkToken(work.getWorkItem().getWorkToken())
+              .setCacheToken(work.getWorkItem().getCacheToken())
+              .build();
+      commits.add(Commit.create(commitRequest, createComputationState("computationId-" + i), work));
+    }
+
+    workCommitter.start();
+    commits.forEach(workCommitter::commit);
+
+    Map<Long, WorkItemCommitRequest> committed =
+        fakeWindmillServer.waitForAndGetCommits(commits.size() / 2);
+
+    for (Commit commit : commits) {
+      if (commit.work().isFailed()) {
```

Comment:
verify that non-failed ones are in committed

Reply:
done